2021/Miglior/main.py

```diff
@@ -0,0 +1,227 @@
+    ds = res[1]
+    rows = res[2]
+
+    print('*** NOTICE: RRD step is {}s, and selected candle step is {} -> each candle will represent {} minutes of data ***\n'.format(
+        step,
+        CANDLE_STEP,
+        step*CANDLE_STEP/60)
+    )
+
+    return start, end, step, ds, rows
+
+
+# Compute some forecast data for candlestick represented time series in pandas dataframe df.
+# each row of the dataset represents a candle in the candlestick, having the following format:
+#
+#            | DateIndex | open value | close value | max value | min value | volume
+#
+# this function will add forecast values forecasting to df, using exponential moving average methods.
+
+
+def forecast_data(df, method, alpha, beta, delta):
+    if method == 'simple':
+        df['open_forecast'] = sm.simple_exponential_smoothing(df['open'], alpha=alpha)
+        df['close_forecast'] = sm.simple_exponential_smoothing(df['close'], alpha=alpha)
+        df['volume_forecast'] = sm.simple_exponential_smoothing(df['volume'], alpha=alpha)
+
+    elif method == 'double':
+        df['open_forecast'] = sm.double_exponential_smoothing(df['open'], alpha=alpha, beta=beta)
+        df['close_forecast'] = sm.double_exponential_smoothing(df['close'], alpha=alpha, beta=beta)
+        df['volume_forecast'] = sm.double_exponential_smoothing(df['volume'], alpha=alpha, beta=beta)
+
+    # compute std and fill df with confidence bands
+    # setting ddof to 0 in order to calculate empirical variance (numpy behaviour)
+
+    open_std = (abs(df['open_forecast'] - df['open'])).std(ddof=0)
+    close_std = (abs(df['close_forecast'] - df['close'])).std(ddof=0)
+    volume_std = (abs(df['volume_forecast'] - df['volume'])).std(ddof=0)
+
+    df['op_confidence_upper_band'] = [val + delta * open_std for val in df['open_forecast']]
+    df['op_confidence_lower_band'] = [val - delta * open_std for val in df['open_forecast']]
+    df['vol_confidence_upper_band'] = [val + delta * volume_std for val in df['volume_forecast']]
+    df['vol_confidence_lower_band'] = [val - delta * volume_std for val in df['volume_forecast']]
+    df['close_confidence_upper_band'] = [val + delta * close_std for val in df['close_forecast']]
+    df['close_confidence_lower_band'] = [val - delta * close_std for val in df['close_forecast']]
+
+
+if __name__ == '__main__':
+
+    # load configuration
+    build_config()
+
+    print('> fetching data from database...')
+
+    start, end, step, ds, rows = fetch_rrd_data(DB)
+    ds_rows = []
+
+    for x in rows:
+        ds_rows.append(x[DS_INDEX])
+
+    # Building candles and creating pandas dataframe to analyze data.
+    candles = build_candles(start, step, ds_rows, CANDLE_STEP)
+
+    # Build pandas dataframe
+    df = pd.DataFrame.from_records(s.to_dict() for s in candles)
+    quantiles = {
+        'volume': df['volume'].quantile(QUANTILE),
+        'open': df['open'].quantile(QUANTILE),
+        'close': df['close'].quantile(QUANTILE)
+    }
```

Comment:
Credo che dovresti mappare questi open/close/volume in un q.cosa che si applichi al mondo delle reti, cosi' come discusso in precedenza.